src/org/pentaho/agilebi/modeler/models/annotations/AnnotationType.java

```diff
@@ -359,6 +375,169 @@ public void populate( final Map<String, Serializable> propertiesMap ) {
     }
   }
 
+  /**
+   * Returns the physical column name that this annotation should operate on. For sources
+   * based on HierarchyLevels we need to consult the existing model to find the underlying
+   * source field.
+   *
+   * @param modelerWorkspace Workspace to look for the level
+   * @param levelName Level formula to find the level
+   * @param cubeName Cube name to file the level
+   * @return field
+   * @throws ModelerException
+   */
+  protected String resolveFieldFromLevel( final ModelerWorkspace modelerWorkspace,
+                                          final String levelName,
+                                          final String cubeName ) throws ModelerException {
+    if ( !StringUtils.isBlank( levelName ) && !StringUtils.isBlank( cubeName ) ) {
+      String locale = Locale.getDefault().toString();
+      LogicalModel businessModel = modelerWorkspace.getLogicalModel( ModelerPerspective.ANALYSIS );
+      List<OlapCube> olapCubes = (List<OlapCube>) businessModel.getProperty( OLAP_CUBES_PROPERTY );
+      OlapCube olapCube = null;
+      for ( int c = 0; c < olapCubes.size(); c++ ) {
+        if ( ( (OlapCube) olapCubes.get( c ) ).getName().equals( cubeName ) ) {
+          olapCube = (OlapCube) olapCubes.get( c );
+          break;
+        }
+      }
+      if ( olapCube == null ) {
+        throw new ModelerException(
+          BaseMessages.getString( "ModelAnnotation.resolveField.UNABLE_TO_FIND_CUBE", cubeName )
+        );
+      }
+
+      List usages = olapCube.getOlapDimensionUsages();
+      for ( int u = 0; u < usages.size(); u++ ) {
+        OlapDimensionUsage usage = (OlapDimensionUsage) usages.get( u );
+        OlapDimension olapDimension = usage.getOlapDimension();
+        List olapHierarchies = olapDimension.getHierarchies();
+        for ( int h = 0; h < olapHierarchies.size(); h++ ) {
+          StringBuffer buffer = new StringBuffer();
+          buffer.append( "[" );
+          buffer.append( usage.getName() );
+          OlapHierarchy olapHierarchy = (OlapHierarchy) olapHierarchies.get( h );
+          if ( StringUtils.isNotEmpty( olapHierarchy.getName() )
+            && !StringUtils.equals( olapHierarchy.getName(), usage.getName() ) ) {
+            buffer.append( "." ).append( olapHierarchy.getName() );
+          }
+          buffer.append( "].[" );
+          List hierarchyLevels = olapHierarchy.getHierarchyLevels();
+          for ( int hl = 0; hl < hierarchyLevels.size(); hl++ ) {
+            OlapHierarchyLevel olapHierarchyLevel = (OlapHierarchyLevel) hierarchyLevels.get( hl );
+            if ( levelName.equals( buffer.toString() + olapHierarchyLevel.getName() + "]" ) ) {
+              return olapHierarchyLevel.getReferenceColumn().getName( locale );
+            }
+          }
+        }
+      }
+      throw new ModelerException(
+        BaseMessages.getString( "ModelAnnotation.resolveField.UNABLE_TO_FIND_LEVEL", levelName )
+      );
+    } else {
+      throw new ModelerException(
```

Comment:
If the else block just throws an exception, why not just test if the fields are blank and throw the exception immediately instead of nesting the rest of the method's logic